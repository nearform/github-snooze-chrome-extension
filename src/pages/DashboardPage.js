import React, { useState } from 'react'
import { Typography, Alert, Divider, Box } from '@mui/material'
import DialogFormButton from '../components/DialogFormButton'
import SnoozeItem from '../components/SnoozeItem'
import { addSnooze } from '../api/chrome'
import { addHours } from '../date'
import { getEntityInfo } from '../parser'
import { SNOOZE_STATUS_PENDING } from '../constants'
import { getEntity } from '../api/github'

function DashboardPage({
  isAuthenticated,
  pat,
  user,
  currentUrl,
  snoozeList: snoozes
}) {
  const [errorMessage, setErrorMessage] = useState('')
  const [snoozeList, setSnoozeList] = useState(snoozes)

  const handleAddSnooze = async hours => {
    setErrorMessage('')

    const numberOfHours = parseFloat(hours)
    if (!numberOfHours || isNaN(numberOfHours)) {
      setErrorMessage(
        'Snooze not added. Please, insert a valid number of hours.'
      )
      return
    }

    const now = new Date()
    const notifyDate = addHours(now, numberOfHours)
    const unixTimestamp = notifyDate.getTime()

    const entityInfo = getEntityInfo(currentUrl)
    const entity = await getEntity(entityInfo, pat)
    const { updated_at: updatedAt } = entity

    const snooze = {
      url: currentUrl,
      notifyAt: unixTimestamp,
      entityInfo: { ...entityInfo, updatedAt },
      status: SNOOZE_STATUS_PENDING
    }

    const updatedSnoozeList = await addSnooze(user.id, snooze)
    setSnoozeList(updatedSnoozeList)
  }

  const handleDelete = updatedSnoozeList => {
    setSnoozeList(updatedSnoozeList)
  }

  if (!isAuthenticated) {
    return (
      <Typography variant="body1" component="p">
        Please LOGIN to use this Chrome Extension.
      </Typography>
    )
  }
  return (
    <>
      <Typography variant="body1" component="p">
        {`Hello ${user.login} (${user.id})`},
      </Typography>
      <Divider />
      {errorMessage && (
        <>
          <Box height={20} />
          <Alert severity="error">{errorMessage}</Alert>
        </>
      )}
      {currentUrl === null && (
        <>
          <Box height={20} />
          <Alert severity="warning">
            This is not a valid URL for adding a GitHub Snooze
          </Alert>
        </>
      )}
      <Box height={20} />
      <DialogFormButton
        label="Add Snooze"
        title="Add Snooze"
        description="In how many hours do you want to be notified for this item?"
        placeholder="Number of hours"
        onConfirm={handleAddSnooze}
        disabled={currentUrl === null}
      />
      <Box height={20} />
      <Divider />
      <Box height={20} />
      <Typography variant="body1" component="p">
        Your Snoozes:
      </Typography>
      {snoozeList.length === 0 && (
        <Typography variant="subtitle2" component="p">
          Snooze List is empty.
        </Typography>
      )}
      {snoozeList.map((snooze, index) => (
        <SnoozeItem
          key={index}
          index={index}
          snooze={snooze}
          onDelete={handleDelete}
        />
      ))}
    </>
  )
}

export default DashboardPage
