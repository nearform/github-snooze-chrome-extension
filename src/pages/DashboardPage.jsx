import React from 'react'
import { Typography, Alert, Divider, Box } from '@mui/material'
import T from 'prop-types'
import SnoozeItem from '../components/SnoozeItem'
import Gap from '../components/Gap'
import { SnoozeShape } from '../shapes'

export default function DashboardPage({
  errorMessage,
  isAuthenticated,
  currentUrl,
  setSnoozeList,
  snoozeList,
  user
}) {
  if (!isAuthenticated) {
    return (
      <Typography variant="body1" component="p">
        Please LOGIN to use this Chrome Extension.
      </Typography>
    )
  }

  return (
    <>
      {errorMessage && <Alert severity="error">{errorMessage}</Alert>}
      {!currentUrl && (
        <>
          <Alert severity="warning">
            The URL you are navigating on is not a valid URL for adding a GitHub
            Snooze
          </Alert>
          <Gap />
          <Divider />
        </>
      )}

      <Gap />
      {!snoozeList.length && (
        <Typography variant="subtitle2" component="p">
          Snooze List is empty.
        </Typography>
      )}
      {snoozeList.map(snooze => (
        <Box key={snooze.id} sx={{ mb: 1.5 }}>
          <SnoozeItem user={user} snooze={snooze} onDelete={setSnoozeList} />
        </Box>
      ))}
    </>
  )
}

DashboardPage.propTypes = {
  snoozeList: T.arrayOf(SnoozeShape).isRequired
}
