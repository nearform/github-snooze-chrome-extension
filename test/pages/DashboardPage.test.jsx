import React from 'react'
import { render } from '../renderer'
import DashboardPage from '../../src/pages/DashboardPage'

const props = {
  isAuthenticated: false,
  pat: '123',
  currentUrl: 'https://example.com',
  snoozeList: [],
  user: {
    id: 123,
    login: 'User'
  }
}

describe('DashboardPage.js', () => {
  test('if not authenticated renders please login', () => {
    const { asFragment } = render(<DashboardPage {...props} />)

    expect(asFragment()).toMatchSnapshot()
  })

  test('if authenticated renders the proper page', async () => {
    chrome.storage.sync.get = jest.fn(() => ({ 123: [] }))

    const { asFragment } = render(
      <DashboardPage {...props} isAuthenticated={true} />
    )

    expect(asFragment()).toMatchSnapshot()
  })
})
