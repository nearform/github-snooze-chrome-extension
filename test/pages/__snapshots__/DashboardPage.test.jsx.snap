// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DashboardPage.js if authenticated renders the proper page 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-1pg8bi"
  />
  <p
    class="MuiTypography-root MuiTypography-subtitle2 css-1ob7xgl-MuiTypography-root"
  >
    Snooze List is empty.
  </p>
</DocumentFragment>
`;

exports[`DashboardPage.js if not authenticated renders please login 1`] = `
<DocumentFragment>
  <p
    class="MuiTypography-root MuiTypography-body1 css-14sfki5-MuiTypography-root"
  >
    Please LOGIN to use this Chrome Extension.
  </p>
</DocumentFragment>
`;
